README.md
# flutter_banner

A Flutter banner widget, only build once after created.

## Getting Started
### 1、add dependency to you `pubspec.yaml`
```
    flutter_banner: ^1.1.0
```

### 2、user guide, see `example/example_main.dart` 
```
    import 'package:flutter_banner/banner_widget.dart'

    BannerWidget banner = BannerWidget(
        delegate: BannerDelegate(
            childAtIndex: (idx) => Container(),
            numberOfBanners: 5,
        ),
        pageIndicator: PageIndicator(),
    );
    banner.delegate.numberOfBanners.value = data.length;
    banner.pageIndicator.numberOfPages.value = data.length;

    return Container(
        child: banner,
        width: 200,
        height: 80,
    );
```

#### `PageIndicator` style
#### ①Replace PageIndicator default style
```
    PageIndicator.normalIndicatorStyle = IndicatorStyle(Size(), BoxDecoration());
    PageIndicator.selectedIndicatorStyle = IndicatorStyle(Size(), BoxDecoration());
```
#### ②Custom `PageIndicator` style
```
    static Widget indicatorBuilder() {
        return Container();
    }

    static Widget selectedIndicatorBuilder() {
        return Container();
    }
    PageIndicator(indicatorBuilder: indicatorBuilder, selectedIndicatorBuilder: selectedIndicatorBuilder,);
```

### 3、support autoloop，user draging

CHANGELOG.md
## [1.0.0] - banner widget 支持手势、自动滚动.
## [1.0.1] - banner widget 暴露停止、启动自动滚动接口.
## [1.1.0] - banner widget 修复banner刷新时异常.
## [1.2.0] - banner widget 修复banner滚动Bad State异常

example/example_main.dart
import 'dart:core';
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

import '../lib/banner_widget.dart';

void main() {
  runApp(MyApp());
}


class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  BannerDelegate _bannerDelegate;
  BannerWidget _bannerWidget;

  @override
  void initState() {
    super.initState();

    List<String> data = ['Page 0', 'Page 1', 'Page 2'];
    _bannerDelegate = BannerDelegate(
      autoLoop: true,
      infinite: true,
      childAtIndex: (int idx) {
        return Text(data[idx],
          textAlign: TextAlign.center,
          style: TextStyle(
            fontSize: 14,
            color: Colors.cyan,
          ),
        );
      },
      numberOfBanners: data.length,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(6),
        color: Colors.white,
      ),
      onTap: (int index) {
        print('onTap $index');
      },
      width: 200,
      height: 60,
    );
    _bannerWidget = BannerWidget(
      delegate: _bannerDelegate,
      onPageChange: (int _old, int _new) {
        print('onPageChange: $_old to $_new');
      },
      pageIndicator: PageIndicator(numberOfPages: data.length),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      backgroundColor: Colors.orangeAccent,
      body: Center(
        child: _bannerWidget,
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/src/banner_widget.dart
import 'dart:async';
import 'package:flutter/material.dart';

import 'banner_delegate.dart';
export 'banner_delegate.dart';

import 'page_indicator_widget.dart';
export 'page_indicator_widget.dart';

/// 实现思路：使用 n+2 个视图，在达到边界时滚动PageView到对应的页面。PageView的page不是真正的页码，页码需要单独计算。
/// 
///   * 边界1：当前页面为第0页，左滑切换到倒数第1页，此时调整PageView到倒数第二个位置，便于继续左滑；
/// 
///   * 边界2：当前索引为倒数第1页，右滑到第0页，此时调整PageView到第1个位置，便于继续右滑。
/// 
/// 其他实现方案：
/// 
///   1、同样采用 n+2 个视图的banner_view[https://github.com/yangxiaoweihn/BannerView]；
/// 
///   2、采用 n*IntegerMax 个视图的flutter_banner_swiper[https://github.com/liuwangle/flutter_banner_swiper]。
class BannerWidget extends StatefulWidget {

  BannerWidget({
    @required this.delegate,
    this.onPageChange,
    this.pageIndicator,
  }): assert(delegate != null);
  
  final BannerDelegate delegate;
  final void Function(int, int) onPageChange;
  /// 单页时隐藏 Indicator
  final PageIndicator pageIndicator;

  // 当前页码
  int get currentPage => _pageNotifier.value;
  final ValueNotifier<int> _pageNotifier = ValueNotifier(0);

  /// 轮播控制
  final ValueNotifier<bool> _stopLoopNotifier = ValueNotifier(false);
 
  /// 停止滚动
  void stopLoop() => _stopLoopNotifier.value = true;
 
  /// 开始滚动
  void startLoop() => _stopLoopNotifier.value = false;
 
  @override
  State<StatefulWidget> createState() => _BannerWidgetState();
}

class _BannerWidgetState extends State<BannerWidget> {
  /// 当前页码
  int _currentPage = 0;
  /// 构建PageView子视图
  List<Widget> _children = List();
  /// 保留最原始的数据，即根据数据源 widget.delegate.numberOfBanners 创建的widget
  List<Widget> _cache = List();
  /// PageView 页码控制
  PageController _pageController;
  /// 页面动画持续时长
  Duration _pagingDuration;
  /// 总页数
  int get _numberOfPages => widget.delegate.numberOfBanners.value;
  /// PageView 滚动页数。总页数请使用 _numberOfPages
  int get _pages => _infinite ? _numberOfPages + 2 : _numberOfPages;
  /// 当前索引，页码请使用_currentPage
  int _index = 0;
  /// _pageController.jumpToPage方法调用，调整PageView位置。
  bool _jumping = false;
  /// 是否无限滚动
  bool get _infinite => widget.delegate.infinite && _numberOfPages > 1;
  /// 是否自动滚动
  bool get _autoLoop => widget.delegate.autoLoop && _numberOfPages > 1;
  /// 自动滚动定时器
  Timer _autoLoopTimer;
  /// 是否用户拖拽
  bool _userDraging = false;
  /// 用户手势检测
  Timer _userGestureDetectorTimer;
  /// 用户手势检测次数计数。目前间隔50毫秒计数一次，累计 _needDetectorUserGestureTimes 次如果没有调用停止滚动，默认开启自动滚动定时器。
  /// 有可能检测 _needDetectorUserGestureTimes 次之后用户手指还在PageView上，但这时已经没有太大意义了。
  int _userGestureDetectorCounter = 0;
  /// 总共检测用户手势的次数
  final int _needDetectorUserGestureTimes = 30;
  /// ScrollController.position Bad State 异常时自动重试次数（默认3次）
  /// 出现异常是因为当前Widget没有attach或者被detach了
  int _retryWhenBadStateExceptionTimes = 0;

  @override
  void initState() {
    super.initState();
    _setup();
    widget._stopLoopNotifier.addListener(_loopControl);
    widget.delegate.numberOfBanners.addListener(_bannerTotalPagesChange);
    if (_autoLoop) {
      _startAutoLoopTimer();
    }
  }

  @override
  void dispose() {
    _pageController.dispose();
    _stopAutoLoopTimer();
    _stopUserGestureDetectorTimer();
    widget._stopLoopNotifier.removeListener(_loopControl);
    widget.delegate.numberOfBanners.removeListener(_bannerTotalPagesChange);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (_numberOfPages == 0)  return Container(height: 0,);

    Widget contentWidget = _contentWidget();
    if (_autoLoop == false)   return contentWidget;

    return NotificationListener(
      child: contentWidget,
      onNotification: _notificationListener,
    );
  }

  /// 初始化数据
  void _setup() {
    _children.clear();
    _cache.clear();
    _currentPage = 0;
    _jumping = false;
    _userDraging = false;
    _pagingDuration = widget.delegate.duration ?? Duration(milliseconds: 250);
    _stopUserGestureDetectorTimer();
    _stopAutoLoopTimer();

    for (int idx = 0; idx < _numberOfPages; idx ++) {
      Widget item = widget.delegate.childAtIndex(idx);
      _cache.add(item);
    }
    _children.addAll(_cache);

    if ((_infinite || _autoLoop) && _cache.length > 1) {
      _children.insert(0, _cache.last);
      _children.add(_cache.first);
      _pageController = PageController(initialPage: 1);
      _index = 1;
    }
    else {
      _pageController = PageController();
      _index = 0;
    }
    widget._pageNotifier.value = _currentPage;
  }

  /// 总页数变化，先停止定时器，然后如果需要再重新开启
  void _bannerTotalPagesChange() {
    _setup();
    if (context != null && mounted) {
      setState(() {});
    }
    if (_autoLoop) {
      _startAutoLoopTimer();
    }
  }

  /// 轮播控制
  void _loopControl() {
    if (widget._stopLoopNotifier.value == true || _numberOfPages < 2) {
      _stopAutoLoopTimer();
    }
    else {
      _startAutoLoopTimer();
    }
  }

  void _startAutoLoopTimer({ bool flag = true }) {
    if (flag != true) return;
    if (_autoLoop == false || _autoLoopTimer != null) return;
    Duration autoLoopInterval = widget.delegate.autoLoopInterval ?? Duration(seconds: 3); 
    _autoLoopTimer = Timer.periodic(autoLoopInterval, (timer) {
      _executePageChange(_animateToNextPage);
    });
  }

  void _stopAutoLoopTimer() {
    if (_autoLoopTimer == null) return;
    _autoLoopTimer?.cancel();
    _autoLoopTimer = null;
  }

  // 开启用户手势检测轮询
  void _startUserGestureDetectorTimer() {
    if (_autoLoop == false || _userGestureDetectorTimer != null)  return;
    _userGestureDetectorTimer = Timer.periodic(Duration(milliseconds: 50), (timer) {
      if (_userDraging == false || _userGestureDetectorCounter >= _needDetectorUserGestureTimes) {
        _userDraging = false;
        _stopUserGestureDetectorTimer();
        _startAutoLoopTimer();
      }
      else {
        _userGestureDetectorCounter ++;
      }
    });
  }

  void _stopUserGestureDetectorTimer() {
    _userGestureDetectorCounter = 0;
    if (_autoLoop == false || _userGestureDetectorTimer == null)  return;
    _userGestureDetectorTimer?.cancel();
    _userGestureDetectorTimer = null;
  }

  // UserScrollNotification 之后可能不会调用 ScrollEndNotification
  // 需要开启轮询检测用户当前是否还在拖拽
  bool _notificationListener(Notification noti) {
    if (noti is UserScrollNotification) {
      if (_userDraging == false) {  // 用户开始触摸
        _userDraging = true;
      }
      _debugLog('_BannerWidgetState.UserDrag : $_userDraging');
      // 用户拖拽需要一直检测手势是否移除，自动翻页定时器一直停止
      _stopUserGestureDetectorTimer();
      _startUserGestureDetectorTimer();
      _stopAutoLoopTimer();
    }
    else if (noti is ScrollStartNotification) { // 调用一次
      _debugLog('_BannerWidgetState.StartScroll: $_userDraging');
    }
    else if (noti is ScrollUpdateNotification) {  // 滚动中多次调用
    }
    else if (noti is ScrollEndNotification) {
      _debugLog('_BannerWidgetState.EndScroll: $_userDraging');
      if (_userDraging) {   // 停止滚动
        _stopUserGestureDetectorTimer();
        _userDraging = false;
        _stopAutoLoopTimer();
        _startAutoLoopTimer();
      }
    }
    return false;
  }

  void _pageChange(int page) {
    if (_infinite == false && _autoLoop == false) {
      _pageChangeNotifier(_currentPage, page);
      return;
    }

    if (_jumping) {  // PageView自身滚动区域调整，这时不需要再次调用 widget.onPageChange。否则会调用多次
      _jumping = false;
      return;
    }
    if (_index == _pages - 2 && page == _pages - 1) {
      // 当前已是最后一页，但仍然向左拖拽（或定时器继续向右滚动）。需要恢复到第0页。
      _jumping = true;
      Future.delayed(_pagingDuration).whenComplete(() {
        _executePageChange(() {
          _animateToFirstPage(page);
        });
      });
    }
    else if (_index == 1 && page == 0) {
      // 当前是第0页，但仍然向右拖拽。需要恢复到最后一页。
      _jumping = true;
      Future.delayed(_pagingDuration).whenComplete(() {
        _executePageChange(() {
          _animateToLastPage(page);
        });
      });
    }
    else {  // 正常滚动、定时器滚动
      _index = page;
      _pageChangeNotifier(_currentPage, page - 1);
      _debugLog('_BannerWidgetState.LoopOrDrug: $_currentPage, idx $_index, page $page');
    }
  }

  /// 页面定时滚动到下一页
  void _animateToNextPage() {
    int page = _index + 1;
    _pageController.animateToPage(page,
      duration: _pagingDuration, 
      curve: widget.delegate.curve ?? Curves.easeInOut,
    );
  }

  /// 页面从最后一页滚动到第一页
  void _animateToFirstPage(int page) {
    _index = 1;
    _pageChangeNotifier(_currentPage, 0);
    _pageController.jumpToPage(_index);
    _debugLog('_BannerWidgetState.DragWhenAtLastPage: $_currentPage, idx $_index, page $page');
  }

  /// 页面从第一页被拖拽到最后一页
  void _animateToLastPage(int page) {
    _index = _pages - 2;
    _pageChangeNotifier(_currentPage, _numberOfPages - 1);
    _pageController.jumpToPage(_index);
    _debugLog('_BannerWidgetState.DragWhenAtFirstPage: $_currentPage, idx $_index, page $page');
  }

  void _pageChangeNotifier(int prePage, int curPage) {
    bool reset = false;
    if (curPage >= _numberOfPages || curPage < 0) { // 页码计算有误
      curPage = 0;
      reset = true;
    }
    _currentPage = curPage;
    widget._pageNotifier.value = _currentPage;
    if (widget.onPageChange != null) {
      widget.onPageChange(prePage, curPage);
    }
    widget.pageIndicator?.updateCurrentPage(_currentPage);
    if (reset == true) {  // 重置
      _bannerTotalPagesChange();
    }
  }

  // 页面滚动时校验 ScrollController、当前 State 状态是否正常
  void _executePageChange(void Function() fn) {
    if (fn == null) return;
    if (mounted == false || _pageController.hasClients == false) {  // 异常时重新开启自动滚动
      _stopAutoLoopTimer();
      _startAutoLoopTimer(flag: _retryWhenBadStateExceptionTimes < 3);  // 重试3次
      _retryWhenBadStateExceptionTimes ++;
      final log = "BannerState: mounted: $mounted, hasClients: ${_pageController?.hasClients}-" +
        "${_pageController?.positions?.toString()}, retry: $_retryWhenBadStateExceptionTimes";
      _debugLog(log);
      if (widget.delegate.exceptionTrace != null) {
        widget.delegate.exceptionTrace(log);
      }
      return;
    }
    _retryWhenBadStateExceptionTimes = 0;
    fn();
  }

  void _debugLog(String log) {
    if (widget.delegate.disableLog) return;
    if (bool.fromEnvironment("dart.vm.product"))  return;
    debugPrint(log);
  }

  Widget _contentWidget() {
    Widget contentWidget = _cache.first;
    if (_numberOfPages > 1) {
      PageView pageView = PageView(
        scrollDirection: widget.delegate.scrollDirection ?? Axis.horizontal,
        children: _children,
        controller: _pageController,
        onPageChanged: _pageChange,
      );
      if (widget.pageIndicator != null) { // 展示 Indicator
        contentWidget = Stack(
          alignment: Alignment.topCenter,
          children: <Widget>[
            Padding(
              padding: widget.delegate.bannerInsets,
              child: pageView,
            ),
            _indicatorWidget(),
          ],
        );
      }
      else {
        contentWidget = pageView;
      }
    }

    return GestureDetector(
      child: Container(
        child: contentWidget,
        decoration: widget.delegate.decoration ?? BoxDecoration(color: Colors.white)
      ),
      onTap: () {
        if (widget.delegate.onTap == null)  return;
        if (_currentPage >= _numberOfPages || _currentPage < 0) { // 页码计算有误
          _bannerTotalPagesChange();    // 模拟重置
          return;
        }
        widget.delegate.onTap(_currentPage);
      },
    );
  }

  Widget _indicatorWidget() {
    return Column(
      children: <Widget>[
        Expanded(child: Container()),
        Padding(
          padding: widget.delegate.indicatorInsets ?? EdgeInsets.only(bottom: 12),
          child: widget.pageIndicator,
        ),
      ],
    );
  }
}

lib/src/banner_delegate.dart
import 'package:flutter/material.dart';

class BannerDelegate {
  BannerDelegate({
    this.autoLoop = false,
    this.infinite = true,
    this.decoration,
    this.onTap,
    this.disableLog = true,
    this.curve = Curves.easeInOut,
    this.duration = const Duration(milliseconds: 250),
    this.autoLoopInterval = const Duration(seconds: 3),
    this.scrollDirection = Axis.horizontal,
    this.bannerInsets = EdgeInsets.zero,
    this.indicatorInsets = const EdgeInsets.only(bottom: 12),
    this.exceptionTrace,
    @required this.childAtIndex,
  }): assert(childAtIndex != null);

  /// banner页数
  final ValueNotifier<int> numberOfBanners = ValueNotifier(0);
  /// 自动循环，设置为ture时忽略infinite属性，默认可无限循环；data长度小于2时无效；
  final bool autoLoop;
  /// 首尾切换效果跟其他页面一致(无限循环)；data长度小于2时无效
  final bool infinite;
  /// 翻页时间
  final Duration duration;
  /// 页面切换动画样式
  final Curve curve;
  /// 翻页间隔时间
  final Duration autoLoopInterval;
  /// 第 idx 位置元素
  final Widget Function(int) childAtIndex;
  /// 点击第 idx 位置元素
  final void Function(int) onTap;
  /// 滚动方向
  final Axis scrollDirection;
  /// banner间距
  final EdgeInsets bannerInsets;
  /// indicator 间距
  final EdgeInsets indicatorInsets;
  /// 背景色
  final BoxDecoration decoration;
  /// 禁用调试日志，默认true
  final bool disableLog;
  /// 异常统计
  final void Function(String) exceptionTrace; 
}

lib/src/page_indicator_widget.dart
import 'package:flutter/material.dart';

class PageIndicator extends StatefulWidget {

  PageIndicator({
    this.indicatorBuilder = PageIndicator.normalStyleBuilder,
    this.selectedIndicatorBuilder = PageIndicator.selectedStyleBuilder,
    this.padding = 10.0,
  }) :  assert(indicatorBuilder != null && selectedIndicatorBuilder != null);

  /// 总共多少页
  final ValueNotifier<int> numberOfPages = ValueNotifier(0);
  /// 每个Indicator的间距
  final double padding;
  /// 正常状态Indicator构建，默认为PageIndicator.builder函数，可自定义
  final Widget Function() indicatorBuilder;
  /// 选中的Indicator
  final Widget Function() selectedIndicatorBuilder;
  /// 页码变化通知，_PageIndicatorState监听页码变化而滚动
  final ValueNotifier<int> _currentPage = ValueNotifier(0);

  // 普通样式 
  static BoxDecoration _normalDecoration = BoxDecoration(
    color: Color(0x33000000), 
    borderRadius: BorderRadius.circular(6.0),
  );
  static IndicatorStyle normalIndicatorStyle = IndicatorStyle(Size(6.0, 6.0), _normalDecoration);
  
  static Widget normalStyleBuilder() {
    IndicatorStyle style = PageIndicator.normalIndicatorStyle;
    return Container(
      decoration: style.decoration,
      width: style.size.width,
      height: style.size.height,
    );
  }

  // 选中样式 
  static BoxDecoration _selectedDecoration = BoxDecoration(
    color: Color(0x99000000),
    borderRadius: BorderRadius.circular(6.0),
  );
  static IndicatorStyle selectedIndicatorStyle = IndicatorStyle(Size(6.0, 6.0), _selectedDecoration);

  static Widget selectedStyleBuilder() {
    IndicatorStyle style = PageIndicator.selectedIndicatorStyle;
    return Container(
      decoration: style.decoration,
      width: style.size.width,
      height: style.size.height,
    );
  }

  /// 更新当前展示的页数
  void updateCurrentPage(int page) => _currentPage.value = page;

  @override
  State<StatefulWidget> createState() => _PageIndicatorState();
}

class _PageIndicatorState extends State<PageIndicator> {
  int _prePage = 0;
  List<Widget> _indicators = List();

  @override
  void initState() {
    super.initState();
    _createDots();
    widget._currentPage.addListener(_reload);
  }

  @override
  void dispose() {
    widget._currentPage.removeListener(_reload);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return RepaintBoundary(
      child: Row(
        children: _children(),
        mainAxisAlignment: MainAxisAlignment.center,
      ),
    );
  }

  // 页面数量没变化时更改page
  void _reload() {
    if (context == null || mounted == false)  return;
    if (_prePage < 0 || _prePage >= widget.numberOfPages.value) return;
    int curPage = widget._currentPage.value;
    if (curPage < 0 || curPage >= widget.numberOfPages.value) return;
    if (_prePage == curPage) return;

    // 交换 上一页(_prePage) 与 当前页(curPage) Indicator 位置
    Widget preWidget = _indicators[_prePage];
    Widget curWidget = _indicators[curPage];
    _indicators[curPage] = preWidget;
    _indicators[_prePage] = curWidget;
    // 更新索引
    _prePage = widget._currentPage.value;

    setState(() { });
  }

  // 创建Indicator圆点
  void _createDots() {
    _indicators.clear();
    for (int idx = 0; idx < widget.numberOfPages.value - 1; idx ++) {
      _indicators.add(widget.indicatorBuilder());
    }
    _indicators.insert(_prePage, widget.selectedIndicatorBuilder());
  }

  List<Widget> _children() {
    List<Widget> items = List();
    _indicators.forEach((indicator) {
      items.add(indicator);
      Container padding = Container(width: widget.padding,);
      items.add(padding);
    });
    items.removeLast(); // 移除最后添加的padding
    return items;
  }
}


class IndicatorStyle {
  IndicatorStyle(this.size, this.decoration)
    : assert(size != null && decoration != null);

  final Size size;
  final BoxDecoration decoration;
}

examples/lib/main.dart
import 'dart:core';
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

import 'package:flutter_banner/banner_widget.dart';

void main() {
  runApp(MyApp());
}


class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  BannerWidget _bannerWidget;

  @override
  void initState() {
    super.initState();

    List<String> data = ['Page 0', 'Page 1', 'Page 2'];
    List<Color> colors = [Colors.amberAccent, Colors.blueAccent];
    _bannerWidget = BannerWidget(
      delegate: BannerDelegate(
        autoLoop: true,
        infinite: true,
        childAtIndex: (int idx) {
          return Container(
            color: colors[idx % 2],
            child: Center(
              child: Text(data[idx],
                textAlign: TextAlign.center,
                style: TextStyle(
                  fontSize: 14,
                  color: Colors.cyan,
                ),
              ),
            )
          );
        },
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(6),
          color: Colors.white,
        ),
        onTap: (int index) {
          print('onTap $index');
        },
      ),
      onPageChange: (int _old, int _new) {
        print('onPageChange: $_old to $_new');
      },
      pageIndicator: PageIndicator(),
    );
    _bannerWidget.delegate.numberOfBanners.value = data.length;
    _bannerWidget.pageIndicator.numberOfPages.value = data.length;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      backgroundColor: Colors.orangeAccent,
      body: Column(
        children: <Widget>[
          Center(
            child: Container(
              child: _bannerWidget,
              width: 200,
              height: 80,
            ),
          ),
          RaisedButton(
            child: Text('stop Loop'),
            onPressed: () {
              _bannerWidget.stopLoop();
            },
          ),
          RaisedButton(
            child: Text('start Loop'),
            onPressed: () {
              _bannerWidget.startLoop();
            },
          ),
        ]
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
